cpp/src/replace/replace.cu

```diff
@@ -19,130 +19,332 @@
+                      bit_mask_t * __restrict__ output_valid,
+                      gdf_size_type *           output_null_count,
+                      gdf_size_type             nrows,
+                      const T* __restrict__ values_to_replace_begin,
+                      const T* __restrict__ values_to_replace_end,
+                      const T* __restrict__           d_replacement_values,
+                      bit_mask_t const * __restrict__ replacement_valid)
   {
-    gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
-    while(i < nrows)
-    {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+  gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+  __shared__ T    temp_data[BLOCK_SIZE];
+
+  while (i < nrows) {
+    bool output_is_valid = true;
+    uint32_t bitmask = 0xffffffff;
+
+    if (input_has_nulls) {
+      bool const input_is_valid{bit_mask::is_valid(input_valid, i)};
+      output_is_valid = input_is_valid;
+
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        temp_data[threadIdx.x] = get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid,
+                                      output_is_valid);
+      }
+
+    } else {
+       temp_data[threadIdx.x] = get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid,
+                                      output_is_valid);
+    }
+
+    __syncthreads(); // waiting for temp_data to be ready
+
+    output_data[i] = temp_data[threadIdx.x];
 
-      if (found_ptr != old_values_end) {
-          auto d = thrust::distance(old_values_begin, found_ptr);
-          d_col_data[i] = d_new_values[d];
+    /* output null counts calculations*/
+    if (input_has_nulls or replacement_has_nulls){
+
+      bitmask &= __ballot_sync(active_mask, output_is_valid);
+
+      // allocating the shared memory for calculating the block_sum and
+      // block_valid_counts
+      __shared__ uint32_t warp_smem[warp_size];
+
+      if(threadIdx.x < warp_size) warp_smem[threadIdx.x] = 0;
+      __syncthreads();
+
+      const int wid = threadIdx.x / warp_size;
+      const int lane = threadIdx.x % warp_size;
+
+      if(lane == 0)warp_smem[wid] = __popc(active_mask);
+      __syncthreads(); // waiting for the sum of each warp to be ready
+
+      uint32_t block_sum = sum_warps<uint32_t>(warp_smem);
```

Comment:
Good observation, but it can be even simpler and faster: get rid of the block sum altogether, and compute the *valid* count rather than the *null* count using the atomicAdd code. Then when the kernel is done, read back the valid count and subtract it from column.size to get null_count.